File: samples/spaceshooter/scripts/intro/state.lua
local milk = require "scripts.milk"
local GameState = require "scripts.state"
local StarField = require "scripts.common.starfield"
local GameplayState = require "scripts.gameplay.state"

local TITLE = "STAR GAME II"
local SUBTITLE = "Return of The Bad Bois"
local PROMPT = "press Z"

local IntroState = GameState:subclass("IntroState")

function IntroState:initialize()
    self.starField = StarField()
end

--  luacheck: push ignore self game
function IntroState:enter(game)
    milk.loadsnd(0, "sounds/intro_music.wav")
    milk.play(0, 0, 128)
end

function IntroState:update(game)
    self.starField:update()

    if milk.btnp(5) then
        game.popState()
        game.pushState(GameplayState())
    end
end

local function printCenteredHorizontal(val, y, scale, color)
    local length = string.len(val)
    local halfLength = length / 2
    local halfResolution = 256 / 2
    local characterSize = 8 * scale

    milk.sprfont(halfResolution - halfLength * characterSize, y, val, scale, color)
end

local function drawTitle()
    printCenteredHorizontal(TITLE, 60, 2, 0x00ff00)
end

local function drawSubtitle()
    local subtitleLength = string.len(SUBTITLE)
    local halfResolution = 256 / 2
    local x = (halfResolution - (subtitleLength / 2) * 8) - 8

    for i = 1, subtitleLength do
        local character = SUBTITLE:sub(i, i)
        local adjustedTime = milk.ticks + i * 4
        local y = 85 + math.sin(adjustedTime / 15) * 4

        milk.sprfont(x + 8 * i, y, character, 1, 0x008751)
    end
end

local function drawPressStart()
    if milk.ticks % 64 < 48 then
        printCenteredHorizontal(PROMPT, 164, 1)
    end
end

function IntroState:draw(game)
    milk.clrs()
    self.starField:draw()
    drawTitle()
    drawSubtitle()
    drawPressStart()
end
-- luacheck: pop

return IntroState

File: sample/scripts/mainMenu/state.ts
import { Stars } from "../common/stars";
import { Game, GameState } from "../game";
import { Menu } from "./menu";
import { TacoStand } from "../common/tacoStand";

const PRESENTED_BY = "It'sgood's"
const TITLE = "STAR GAME";
const SUBTITLE = "Return of The Bad Boiz";

export class MainMenuState implements GameState {

    public updateBelow = false;
    public drawBelow = false;

    private _starField = new Stars(0.1);
    private _tacoStand = new TacoStand();
    private _menu = new Menu();

    public enter(_: Game): void {
        loadspr("art/sprsheet.bmp");
        openstream(0, "sounds/12 HHavok-main.wav");
        playstream(0, 128, true);
    }

    public update(game: Game): void {
        this._starField.update();
        this._tacoStand.update(game.ticks);
        this._menu.update(game);
    }

    public draw(game: Game): void {
        clrs(0x00);

        this._starField.draw();
        this._tacoStand.draw();
        this._menu.draw(256 / 2, 150, game.ticks);

        this.drawTitle();
        this.drawSubtitle(game.ticks);
    }

    public exit(_: Game): void {
        closestream(0);
    }

    private drawTitle(): void {
        const length = string.len(TITLE);
        const halfLength = length / 2;
        const halfResolutionWidth = 256 / 2;

        font(0, halfResolutionWidth - halfLength * 16, 45, PRESENTED_BY, 1, 0x00ffffff);
        font(0, halfResolutionWidth - halfLength * 16, 60, TITLE, 2, 0x00ff000);
    }

    private drawSubtitle(ticks: number): void {
        const length = string.len(SUBTITLE);
        const halfLength = length / 2;
        const halfResolutionWidth = 256 / 2;
        const x = halfResolutionWidth - halfLength * 8;

        for (let i = 1; i <= length; i++) {
            const character = string.sub(SUBTITLE, i, i);
            const adjustedTime = ticks + i * 4;
            const y = 85 + math.sin(adjustedTime / 15) * 4;

            font(0, x + 8 * i, y, character, 1, 0x008751);
        }
    }
}
